### DISCUSSION TITLE
workflow with concurrency

### DISCUSSION THREAD

**Discussion by Rageshmookoth** [community/community#33142] [Actions]
I have  6 workflows where requirement is to run in sequential fashion, no  dependency . however at given time only one workflow out of 
 6 should be running. 

I used concurrency as below.  mentioned same group for all the 6 workflow.  It works, however it triggering as 1 in progress,2 queued jobs  and cancelling remaining workflow.   

Expected  result is 9 jobs in queue and 1 in progress.   Can someone please help me 

concurrency:
  group:  test
  cancel-in-progress: false

**Comment by SPodjasek** [comment]
IMHO it is currently not possible. According to docs, any pending jobs/workflows in the same concurrency group will get canceled.
I'm facing the same issue because I want to limit my workflows (started with `workflow_dispatch`) to run sequentially, one at a time, and as far as I've found, it is impossible.
Maybe some suggestion to GitHub to implement such a feature to limit running instances of workflow without canceling any pending runs, keep them pending until the currently running one finishes.

**Reply by SPodjasek** [reply]
There's a possible workaround to this issue, but it can accrue additional costs. You can use either self-hosted or new (beta) [GitHub-hosted larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners).
For a self-hosted runner, you create just a single instance runner, and it will pick one job at-a-time. 
For GitHub-hosted, you create a new runner, and in "Auto-scaling," limit the maximum runners to 1.
But as I've mentioned, larger runners aren't included in your plan and are billed by the minute.

**Comment by jokreliable** [comment]
concurrency is for allowing your jobs to shortcut when, for example, you have lots of pushes that trigger workflow from a branch, but only care about the latest results, not the one in the middle... (say your workflow takes 10 minutes, and you push every 2 mins a set of small changes... without concurrency, you'll have a new set of actions triggered for each push. With concurrency, you can abort already running and/or queued jobs that would do the same thing on obsolete refs to not waste time on them, and getting to the final run you need much faster).

Now, the easiest way would be to combine those jobs in one large workflow, so you can setup the proper dependencies between each job, so they can run sequentially (using `needs` on each job). If you don't want to (need the workflows to be managed/approved by different groups or something for example, you can try to setup something like this:

Assuming the following structure: each workflow (say a through f) contains 3 jobs (1,2,3).
Update workflow b through f to be manually triggered (your workflow dispatch should already allow that)
Update workflow a with job 4, that needs completion of all 3 jobs
That job likely does some checks, and then triggers workflow b (providing the necessary parameters to the workflow itself as needed)
Update workflow b through e the same way, triggering the "next workflow" similarly.

Now, when the "original trigger" happens (whatever the event, user/push/pr/etc trigger), only the first workflow is started. this will create jobs 1 2 3 with the dependencies defined there. Once they are all complete, job 4 (which depends on all 3) is then run, which triggers the workblow-b. that then creates the jobs 1,2,3 from workflow b, etc...

There will be interesting corner cases to explore when you cancel things mid-way through (either through the concurrency and/or through the UI by hand). For example, if you have concurrency set, but the run is past the first workflow, you'll end up in a position where it won't be cancelled those workflow-b though f until the new workflow catches up in the chain and cancel the already in progress one. There is likely things you can do to cancel all the possible actions related to that branch/ref/whatever (or previously created jobs).

**Reply by SPodjasek** [reply]
IMHO there should also be an option to force jobs run in series, eg. disable concurrency, but don't cancel any jobs.
As I've mentioned before, when you use actions as CI/CD in GitOps you want every update sent via `workflow_dispatch` to be applied and at the time avoid some rare race conditions when two jobs try to update your repository at the same time.
